### DISCUSSION TITLE
How to move a channel to another category and sort this category by name

### DISCUSSION THREAD

**Discussion by deso85** [discord-jda/JDA#2435] [Questions and Help] [answered]
Hello everyone,

I want to move a channel to a different category and sort all channels inside this category by name. I am able to do that seperately but it won't work if I try to combine these tasks.

Moving the channel:
```
guild.modifyTextChannelPositions()
	.selectPosition(chan)
	.setCategory(cat)
	.complete();
```

Sorting the Category:
```
cat.modifyTextChannelPositions()
	.sortOrder(new ChannelComparator())
	.queue();
```

With ChannelComparator():
```
public class ChannelComparator implements Comparator<GuildChannel> {
	@Override
	public int compare(GuildChannel first, GuildChannel second) {
		if(first == second) {
			return 0;
		}
		if(first == null) {
			return -1;
		}
		if(second == null) {
			return 1;
		}
		
		return first.getName().compareTo(second.getName());
	}
}
```

If I sort after moving the channel the categorys channel list doesn't contain the new channel. I was able to work around that by creating a thread, wait a few seconds and then sort inside that thread. If I just use a Thread.sleep() the categorys channel list won't be updated.

What do I miss? Can I force the categorys channel list to be updated? Is there another way?

Thanks in advance!

**Comment by MinnDevelopment** [comment, marked as answer]
This might work:

```java
channel.getManager().setParent(cat).delay(500, TimeUnit.MILLISECONDS).queue(v -> {
  cat.modifyTextChannelPositions().sortOrder(Comparator.comparing(GuildChannel::getName)).queue();
});
```

1. `setParent(cat)` moves the channel into the category
2. `delay(500, MILLISECONDS)` makes the callback run half a second later (to avoid race conditions)
3. `sortOrder(Comparator.comparing(GuildChannel::getName))` sorts by name

**Reply by deso85** [reply]
It works. Thanks a lot!

Why does the lamda expression work that different?

**Reply by MinnDevelopment** [reply]
Because the thread you are handling events on is the same thread that will handle the cache update. By using `complete()` you prevent the thread from doing other work.
